src/transformers/optimization.py

```diff
@@ -315,3 +315,204 @@ def step(self, closure: Callable = None):
                     p.data.add_(p.data, alpha=-group["lr"] * group["weight_decay"])
 
         return loss
+
+
+# AdaFactor pytorch implementation from FAIR-seq:
+# https://github.com/pytorch/fairseq/blob/master/fairseq/optim/adafactor.py
+#
+# This implementation handles low-precision (FP16, bfloat) values, but we have not thoroughly tested.
+#
+# AdaFactor was introduced by Shazeer at al as a more memory efficient drop-in replacement for Adam,
+# specifically for training large Transformer models.
+# https://arxiv.org/abs/1804.04235
+#
+# For T5 finetuning, recommended settings:
+#   * scheduled LR warm-up to fixed LR, disable relative updates, use clip threshold
+# https://arxiv.org/abs/2004.14546
+#
+# Alternatively, relative_step with warmup_init can also be used.
+# Training without LR warmup or clip threshold, is not recommended. Additional optimizer operations
+# like gradient clipping, should not be used.
```

Comment:
Even though this "second docstring" breaks style convention, I am OK to leave it here, because it is very useful.
It would be better to consolidate with the class docstring below.

Reply:
Gotcha. It's up to you. Happy to move it, or if you want to consolidate the docstring in a future PR. 

Let me try to make the change and see if you like it.